API and web socket updates.

kbenne opened the pull request:
This PR introduces new APIs for:

`/tests` to show running tests
`/testcases/<namespace>` to put test cases into namespaces
`/users/<username>/testcases` for user specific test cases 

Also a new WebSocket API. See tests https://github.com/NREL/boptest-service/blob/issue/24-websocket/test/test_websocket_api.py

jmythms committed:
Create get_tests.md

kbenne committed:
Add documentation for object-storage (Minio) keys

ref #9

kbenne committed:
Reorganize routers into "public" and "authorized"

ref #9

kbenne committed:
Add "authorizer" middleware to fetch user info

kbenne committed:
Avoid passing s3 and sqs through function calls

kbenne committed:
Update docker ignore files for faster builds

kbenne committed:
Add my/testcases routes

ref #9

Commit:
add nodemon to web

kbenne committed:
WIP: Update provision script

jmythms committed:
Create Redis.MD

Commit:
Merge branch 'get_all_tests' of https://github.com/NREL/boptest-service into get_all_tests

Commit:
Merge branch 'develop' into get_all_tests

jmythms committed:
Add Test control points info

jmythms committed:
Add new structure

kbenne committed:
Update provision library

Commit:
Change select

jmythms committed:
REDIS public test routes fixed

jmythms committed:
Merge branch 'issue/9-user-test-cases' of https://github.com/NREL/boptest-service into issue/9-user-test-cases

jmythms committed:
Add /tests

kbenne committed:
Update provision tool with support for user test cases

jmythms committed:
setStatus: fix args + remove unnecessary code

jmythms committed:
fix cleanup on worker side

jmythms committed:
not validated stop

jmythms committed:
working my Tests

jmythms committed:
fix _getTests to work without userID

jmythms committed:
working testid validator

jmythms committed:
fix tests_string output

jmythms committed:
Merge pull request #21 from NREL/public_tests_debug

Pull in changes from public debug

kbenne committed:
Create tests for test case management

kbenne committed:
Add tests for user defined test case api

kbenne committed:
Add configuration for test accounts

This change enables testing without having the dashboard available

kbenne committed:
Simplify config for s3 url

kbenne committed:
Reorganize web src files

kbenne committed:
Merge remote-tracking branch 'origin/develop' into issue/9-user-test-cases

kbenne committed:
Update dockerignore

kbenne committed:
Merge remote-tracking branch 'origin/get_all_tests' into issue/9-user-test-cases

kbenne committed:
Add unit test for get /tests api

kbenne committed:
User user "subject" instead of username

"sub" in the OAuth sense is immutable, whereas BOPTEST allows username
to be modified, therefore it is a not a good key for storage

kbenne committed:
Reorganize and simplify web/server files

kbenne committed:
Remove validation of testid

Validating the testid has become more expensive now that tests can be
associated with a user (which requires a trip to validate the user),
therefore testid validation "in the loop" of simulation apis has been
removed. Now instead of getting an immediate error, the response will
timeout after not receiving a response from a running test.

kbenne committed:
Add feature to abort queued tests

This also required a little rework on the /tests api implementation.
See `test_async_select_api` in test_testcase_api for an example.

kbenne committed:
Update api to return 400 for invalid testid

This is true for all APIs that accept testid as a url param,
but most especially the status api.

kbenne committed:
Use sscan to get all tests

This replaces smembers which is not advised in a production environment

kbenne committed:
Stop tests if they timeout during select

kbenne committed:
Fix warning about `body-parser deprecated undefined extended`

kbenne committed:
Add a default api key for provisioning in dev mode

kbenne committed:
Update readme

kbenne committed:
Fix version api regression

kbenne committed:
Update test environment and correct failing tests

kbenne committed:
Update dockerignore

kbenne committed:
Update redis.md

kbenne committed:
Update Redis.MD

kbenne committed:
Add WebSocket API

Tests and examples still to come.

kbenne committed:
Add websocket example for running one boptest

kbenne committed:
Add test to demonstrate advancing multiple tests over websocket

kbenne committed:
Adjust default number of parallel boptests in unit test

kbenne committed:
Bump to v0.2.0

kbenne committed:
Align pub/sub message format with canonical BOPTEST api doc

Previously there was some light transformation of the data sent over
pub/sub, resulting in a slight difference between the api payloads documented
on the IBPSA BOPTEST repo and the expected pub/sub payloads.

This has not been an issue in the past because the differences were
internal, and the rest API of BOPTEST-Service was still aligned with the
canonical BOPTEST. With the introduction of a websocket api, clients are
given access to directly put messages on the pub/sub bus, so for the
sake of clarity it is good to have alignment.

kbenne committed:
Improve error messages (especially for websocket)

kbenne commented:
@tijcolem the environment has changed so we'll need to talk about config changes. I'd like to deploy this to dev for testing.

kbenne commented:
Also, @tijcolem, we might need some nginx changes to proxy the websocket. https://www.nginx.com/blog/websocket-nginx/

kbenne commented:
I need to update the GitHub action to get it passing. The provision script changed a little.

kbenne committed:
Adjust github action

kbenne committed:
Update github action

kbenne committed:
Run boptest-service test as an action

kbenne committed:
Use check instead of assert in tests

kbenne committed:
Update GitHub action

kbenne commented:
This PR will require a coordinated deployment with the dashboard v1.2 tag https://github.com/NREL/boptest-dashboard/releases/tag/v1.2

tijcolem commented on test/test_websocket_api.py:
```diff
@@ -0,0 +1,146 @@
+import uuid
+from pytest_check import check
+
+host = os.environ.get("BOPTEST_SERVER", "http://web")
```
We may want to change BOPTEST_SERVER to just the domain name? I noticed the websockets.connect("ws://web") is currently hard coded and uses different protocol "ws" vs "http" 

BOPTEST_SERVER=api.dev.boptest.net

kbenne replied:
how about BOPTEST_SERVER and BOPTEST_WS_SERVER so that we can config http versus https?

tijcolem replied:
right, forgot about HTTPS.  I like those two envs

kbenne committed:
Move WebSocket to dedicate endpoint at ws://<host>/ws

kbenne committed:
Send more descriptive error messages on internal server errors

kbenne committed:
Improve stop implementation

Now the test record will be removed from redis right away, meaning that
even if the message to the worker fails (perhaps
because the worker vanished) the test will still be gone from the client
perspective. Even if for some reason the worker is still active for the
test, the idle timeout will clean it up.

kbenne committed:
Return a 401 when trying to select an invalid test case

kbenne committed:
Add a websocket test that sends params

dhblum committed:
Use python arrays instead of list

kbenne committed:
Test overwritting signals via websocket

kbenne committed:
Update docs
